_posts/2018-07-17-its-a-lot-more-snappy.md

```diff
@@ -0,0 +1,174 @@
+
+                if (sizeDifference < this.partitionSize) {
+                    partitionEnd += sizeDifference;
+                } else {
+                    partitionEnd += this.partitionSize;
+                }
+
+                const partition = this.pokemon.slice(partitionStart, partitionEnd);
+                await this.table.bulkPut(partition);
+                partitionStart = partitionEnd;
+            }
+        });
+    }
+}
+```
+
+Now that we had implemented this pattern we noticed that the initial download of the data was taking much longer than it had previously, as we were still downloading the dataset in the same way, performing a mapping operation on the dataset before storing it in our redux store and then inserting all of our newly obtained Pokemon into our Pokemon DB. If only there existed a method, which could enable us to parallelise these operations.
+
+It was at this point our fearless tech lead, William Sia(TM) proposed using a web worker for that exact purpose. This was new and very exciting. Of course we wanted to use a web worker!
+
+I happened to come across a module called `worker-loader` which seamlessly integrated with our current build and packaging eco-system. In fact, worker-loader almost made it inconceivable that we waited such a long time to implement our first web worker. We didn’t even need to add any additional config to our web pack config files.
+
+When using a web worker, all that's needed to get things started is:
+
+- initialise the worker
+- post a message to the worker (preferably as a string)
+- create a callback for when the worker completes
+
+_from the main thread_
+
+```typescript
+function spawnWorker(): void {
+    const bulkPutWorker = new Worker();
+    bulkPutWorker.postMessage(JSON.stringify({ pokemon: this.pokemon, location: this.location }));
+    pokemonActions.startInsertingPokemon();
+    window.addEventListener('pokemonUpdated', pokemonActions.endInsertingPokemon);
+    bulkPutWorker.onmessage = (event: MessageEvent): void => {
+        window.dispatchEvent(new Event(event.data));
+    };
+}
+```
+
+_within the worker_
+
+```typescript
+worker.onmessage = async (event: MessageEvent): Promise<void> => {
+
+    const { location, pokemon } = JSON.parse(event.data) as BulkPutEvent;
+    const flocsUpdater = new FlocsUpdater(location, pokemon);
+    try {
+        await pokemonUpdater.updatePokemon();
+    } catch (error) {
+        console.log('Something has gone awry', error);
+    }
+    worker.postMessage('pokemonUpdated');
+    self.close();
+};
+```
+
+Once we had implemented all of the aforementioned steps, we noticed a dramatic increase in the performance of the application. We no longer had a gradual decrease in performance over time when using the app and finding a specific Pokemon in the IndexedDB table was now an instantaneous operation. Having noticed these improvements, our PM said “After testing the app, it’s now a lot more snappy!”
```

Comment:
`dramatic increase` - lose `dramatic` - weaselwords

Better - give some numbers:  
* Reduced download time by x% 
* Y completed in under Zms
* Consistent responsivness over xHrs use (indicates that previously there was a memory leak, or similar)
* Memory usage dropped by x%

Choose what you want to report based on the message you want to send